[FLINK-2076] [runtime] Fix memory leakage in MutableHashTable

chiwanpark opened the pull request:
Hi. This PR contains a bug fix for [FLINK-2076](https://issues.apache.org/jira/browse/FLINK-2076).

When `prepareNextPartition` method runs with some pending partitions, the memory leakage can occur at `memory.add(getNextBuffer())` statement (Line 515, 516). In common case, there are extra memory spaces for hash table so first and second call of `getNextBuffer()` return memory segment.

But in extreme case such as a case described in JIRA, the second call of `getNextBuffer()` can return null and this null makes the memory leakage.

StephanEwen commented:
Thanks, this looks like some seriously great debugging! Very nice :-)

It would be great if you could add a test that produces the error without the fix, and validates that the fix resolved it. I would guess that you have a setup that produced this error (for debugging). Can you add this as a test?

Also, can we change the fix such that it adds a second memory segment, if it is non-null? That would help maintain the performance characteristics of the current code. I vaguely remember that there was a reason to add two memory segments (that code was written quite a while ago and I should have put more comments into the code).

chiwanpark commented:
I was debugging with ConnectedComponents example and found the bug caused by null memory segment. I tried adding a test case to test re-openable hash table with small memory a moment ago. But I found another memory leakage not related to null memory segment. (All memory segments are not null)

I think that more investigation is needed. I'll update this PR when I fix the bug.

StephanEwen commented:
Awesome, thanks a lot @chiwanpark

chiwanpark commented:
I fixed the bug related to non-null memory segment. The bug was caused by non-clearing `writeBehindBuffersAvailable` variable in `close()` method of `MutableHashTable` class. I added a test case for this bug.

I have a problem for creating a test case to test the bug related to null memory segment. I think that it is related to parallelism also. If I run ConnectedComponents example with parallelism 1, the bug not occurs. So I cannot reproduce the state without ConnectedComponents example.

StephanEwen commented:
Looks good, thanks a lot!

Will merge this...
